tt_eager/tensor/tensor.cpp

```diff
@@ -146,49 +159,72 @@ void Tensor::deallocate(bool force) {
+    // Stall until this worker is done
+    while (true) {
+        std::scoped_lock<std::mutex> lock(this->tensor_attributes->populated_mutex);
+        if (this->tensor_attributes->tensor_populated.at(0)) break;
+    };
 }
 
-void Tensor::wait_for_metadata_populated() const {
-    while (not this->metadata_populated());
+// Worker Thread - Set populated flag to true, once worker has completed it's task for this tensor
+void Tensor::set_populated(Device* worker) {
+    // If worker is not specified, set entry for all workers to true
+    std::scoped_lock<std::mutex> lock(this->tensor_attributes->populated_mutex);
+    if (not worker) {
+        for (int i = 0; i < this->tensor_attributes->tensor_populated.size(); i++) {
+            this->tensor_attributes->tensor_populated.at(i) = true;
+        }
+    }
+    else {
+        this->tensor_attributes->tensor_populated.at(worker->id()) = true;
+    }
 }
 
 void Tensor::deepcopy(const Tensor& other) {
     // Wait until the tensor being copied is populated
-    other.wait_for_metadata_populated();
+    other.wait_for_tensor_data_populated();
     // Populate tensor metadata
     this->set_shape(other.get_shape());
     this->set_storage(other.get_storage());
     this->set_dtype(other.get_dtype());
     this->set_layout(other.get_layout());
     // Set metadata populated flag for getters
-    this->set_metadata_populated();
+    this->set_populated();
 }
 
 void Tensor::populate_buffers_and_metadata(const Tensor& other) {
     // Similar to deepcopy, but to be applied on a tensor that has an empty storage
     // container initialized. Require tensor storage to be correctly initialized.
+    this->set_shape(other.get_shape());
+    this->set_dtype(other.get_dtype());
+    this->set_layout(other.get_layout());
     // Populate storage container with buffers + shapes
     std::visit([this] (auto&& storage) {
         using StorageType = std::decay_t<decltype(storage)>;
         if constexpr(std::is_same_v<StorageType, OwnedStorage> or std::is_same_v<StorageType, DeviceStorage>) {
             std::get<StorageType>(this->tensor_attributes->storage).buffer = storage.buffer;
+            this->tensor_attributes->tensor_populated = {true};
         } else if constexpr(std::is_same_v<StorageType, MultiDeviceHostStorage> or std::is_same_v<StorageType, MultiDeviceStorage>) {
             std::get<StorageType>(this->tensor_attributes->storage).buffers = storage.buffers;
             std::get<StorageType>(this->tensor_attributes->storage).shapes = storage.shapes;
+            this->tensor_attributes->tensor_populated = std::vector<bool>(storage.buffers.size(), true);
         }
     }, other.get_storage()); // Non blocking storage query, since this is done for tensors that get created inside the worker thread
-    // Populate remaining MD
-    this->set_shape(other.get_shape());
-    this->set_dtype(other.get_dtype());
-    this->set_layout(other.get_layout());
-    this->set_metadata_populated();
```

Comment:
How come we don't `set_metadata_populated` for this function? can it be merged with `deepcopy` implementation?